### ISSUE TITLE
TagWith using formattable string takes disproportionately much time

### ISSUE THREAD

**Issue by GertArnold** [dotnet/efcore#32622]
When using a formattable string in `TagWith`, performance drops dramatically.

This code that queries a table in just some table a 1000 times demonstrates the effect of formattable strings:

```c#
var ids = Plots.Take(1000).Select(p => p.PlotId).ToList();

var sw = Stopwatch.StartNew();
foreach (var id in ids)
{
    var plot = Plots.TagWith("Plots selected plot xxxxxx").First(p => p.PlotId == id);
}
var e1 = sw.Elapsed;

sw.Restart();
foreach (var id in ids)
{
    var plot = Plots.TagWith($"Plots selected plot {id}").First(p => p.PlotId == id);
}
var e2 = sw.Elapsed;

sw.Restart();
foreach (var id in ids)
{
    var plot = Plots.TagWith($"Plots selected plot {id}").First(p => p.PlotId == id);
}
var e3 = sw.Elapsed;

sw.Restart();
foreach (var id in ids)
{
    var plot = Plots.TagWith("Plots selected plot xxxxxx").First(p => p.PlotId == id);
}
var e4 = sw.Elapsed;

e1.Dump("No tags");
e2.Dump("Tags");
e3.Dump("Tags");
e4.Dump("No tags");
```

The result:

```
No tags
00:00:00.2720288
Tags
00:00:04.6038496
Tags
00:00:05.9300289
No tags
00:00:00.2160600
```

The result when only generating the strings used in the tags:

```
No tags
00:00:00.0000031
Tags
00:00:00.0000509
Tags
00:00:00.0000499
No tags
00:00:00.0000029
```

Clearly the time added in queries with formattable strings in tags is far more than the overhead of string interpolation.

Strangely enough, first creating the string (`var s = $"Plots selected plot {id}".ToString();`) and putting it in the tag does not make a difference. What's going on here?

FYI: `PlotId` is an `int`, the highest value here is `134647`.

EF Core version: 8.0
Database provider: Microsoft.EntityFrameworkCore.SqlServer 2019
Target framework: NET 6.0
Operating system: win11
IDE: Visual Studio 2022/Linqpad 7

**Comment by roji** [comment]
@GertArnold simple StopWatch benchmarking is almost never a good way to benchmark; always use [BenchmarkDotNet](https://benchmarkdotnet.org/index.html) when measuring speed.

Specifically in your example above... When EF first executes a new LINQ query (and a different TagWith means a new LINQ query), it must go through a lot of work, "compiling" the new query. Once the first execution completes, that LINQ query is fully cached, and subsequent executions should be very fast.

In other words, the increased times you're seeing are very likely only valid for the very first execution of each TagWith variant. Try doing a single execution of each variant outside your StopWatch loop, and you should see fast running times. If you use BenchmarkDotNet, it automatically does "warmup" executions, which aren't counted in the actual running time, precisely to avoid measurement skews like this.

If the above isn't right and you're still seeing slow execution, please post back and I can investigate further.

**Comment by ajcvickers** [comment]
@roji Here you go:

```
| Method       | Mean       | Error    | StdDev   |
|------------- |-----------:|---------:|---------:|
| NoTags       |   154.1 us |  1.36 us |  1.27 us |
| ConstantTags |   164.8 us |  3.00 us |  2.95 us |
| Tags         | 1,876.8 us | 36.99 us | 71.27 us |
```

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using Microsoft.EntityFrameworkCore;

var summary = BenchmarkRunner.Run<Benchmarks>();

public class Benchmarks
{
    private MyContext Context { get; set; } = null!;

    [GlobalSetup]
    public void Setup()
    {
        Context = new MyContext();
        Context.Database.EnsureDeleted();
        Context.Database.EnsureCreated();
    }

    [GlobalCleanup]
    public void Cleanup()
    {
        Context.Dispose();
    }
    
    [Benchmark]
    public void NoTags()
    {
        _ = Context.Plots.FirstOrDefault(p => p.PlotId == _id);
    }

    [Benchmark]
    public void ConstantTags()
    {
        _ = Context.Plots.TagWith("Plots selected plot xxxxxx").FirstOrDefault(p => p.PlotId == _id);
    }

    private static int _id = 0; 
    [Benchmark]
    public void Tags()
    {
        _ = Context.Plots.TagWith($"Plots selected plot {_id++}").FirstOrDefault(p => p.PlotId == _id);
    }
}

public class MyContext : DbContext
{
    public DbSet<Plot> Plots { get; set; }

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        => optionsBuilder
            .UseSqlServer(@"Data Source=localhost;Database=One;Integrated Security=True;TrustServerCertificate=true");
    // .LogTo(Console.WriteLine, LogLevel.Information)
    // .EnableSensitiveDataLogging();
}

public class Plot
{
    public int PlotId { get; set; }
    public string? A { get; set; }
    public string? B { get; set; }
    public required string C { get; set; }
}
```

**Comment by stevendarby** [comment]
@ajcvickers and what if you don't increment _id?

**Comment by ajcvickers** [comment]
@stevendarby Then it's the same as the other two. The issue here is that the continually changing tag is impacting query caching at some level.

**Comment by stevendarby** [comment]
Yeah, isn’t that expected? From @roji’s post:

> a different TagWith means a new LINQ query

> the increased times you're seeing are very likely only valid for the very first execution of each TagWith variant

**Comment by ajcvickers** [comment]
@stevendarby Agreed, but ideally we would recognize that only that tag has changed and hit the cache appropriately.

**Comment by GertArnold** [comment]
Yeah, I was aware of the weaknesses of stopwatch benchmarking. I tried to circumvent it by the specific call order. The reason for investigating it was a considerable performance hit in application code when optimizing a batch process and using `TagWith` to check generated queries in SQL Profiler.
Seeing the comments above it seems to me that the `TagWith` part should be completely ignored when caching queries, if possible?

**Comment by roji** [comment]
@ajcvickers right.

First, the thing is that we integrate the actual tag into the SQL, rather than parameterizing it; this is of course by design: we want it to be a comment that's actually in the SQL so it's e.g. visible in the logs (where parameter values don't get logged etc.). This is also why the [TagWith](https://github.com/dotnet/efcore/blob/main/src/EFCore/Extensions/EntityFrameworkQueryableExtensions.cs#L2805) method parameter has the [NotParameterized] attribute on it, to make sure it's always integrated as a constant into the query tree.

We indeed know that the tag doesn't affect anything in the compilation phase (e.g. same shaper). However, since the SQL must vary for different tag values, we must go through the 2nd part of the query pipeline for different tag values. Since our architecture (currently) only allows caching SQL based on parameter nullability, we'd have to disable caching entirely for this case, i.e. go through the 2nd part of the query again and again each time a query is executed with a parameterized tag.

Concretely, that would mean removing [NotParameterized] on TagWith (to allow ParameterExtractingExpressionVisitor to extract the argument as a parameter instead of make it a constant), and at some point in the 2nd part of the query pipeline (e.g. SqlNullabilityProcessor), recognize TagWith with a parameter argument, change it into a constant, and disable SQL caching (call DoNotCache()).

That would all mean a perf hit (regression) for executing queries with TagWith that has a parameterized argument which **doesn't** change. Though it could be claimed that if the user is parameterizing the TagWith argument, they're doing so in order to change it - and the perf hit is far smaller than the current perf hit of rerunning the entire query pipeline for different values (due to the constantization).

**Comment by roji** [comment]
@GertArnold just to be clear, assuming you have a handful of tags, the performance is completely fine - the first query execution for each tag value would be slow, but subsequent executions would be very fast. If, however, you're trying to have a totally open range of dynamically-generated tag values, then you shouldn't be doing that (with the current architecture). I'd be interested in your actual use-case for that.

**Comment by ajcvickers** [comment]
@roji Right, so it may not be worth doing, but ideally, for parameterized tags, we shouldn't generate the entire SQL again each time. I'm fine with saying we will never do this.

**Comment by GertArnold** [comment]
@roji _I'd be interested in your actual use-case for that._

Just following progress in SQL profiler. See which queries belong together for units of work in a batch process and finding recurring queries as the most promising targets for performance tuning.

Remember, from my POV there was nothing wrong with it, it seemed a perfectly sensible thing to do. It took me quite some time before finally suspecting such an innocent-looking little method!

If it shouldn't be done, fine, I could live with that, but since it's such a natural thing to do there should somehow be an unmissable warning against it.

**Comment by roji** [comment]
@GertArnold my question wasn't about your use-case for using TagWith - that's clear. The question is whether you actually need to use totally dynamic tag values (which are generally going to be different every single time), and if so, why. As I wrote above, if you're using just a single tag value, or several fixed values, there's no performance problem.

**Comment by GertArnold** [comment]
@roji Sorry for not being clear enough, but I tried to explain it by "See which queries belong together for units of work in a batch process". I.e. each repetitive group if queries is easily distinguished by the same moniker and adding a number to the tag gives an impression of the batch's progress. Once knowing the consequences, I can easily do without, but my point it that it's not at all obvious that such dynamic tags should be avoided (principle of least surprise).

**Comment by roji** [comment]
@GertArnold thanks.

**Comment by joakimriedel** [comment]
>  The question is whether you actually need to use totally dynamic tag values (which are generally going to be different every single time), and if so, why. As I wrote above, if you're using just a single tag value, or several fixed values, there's no performance problem.

Interesting. I just stumbled upon this thread, and realized I've been using parameterized TagWith for some time and now I need to get in the codebase and clean all up, I have noticed slowdowns but would never have realized it is because of tagging! 

My use case was to surface details into the Application Insights error log where it won't show parameter values, but the TagWith comment is visible. I started adding a lot of dynamic variables (batch numbers etc) to aid debugging whenever I get issues with an Azure App service.

Time to rewrite all TagWith to constant values 😕

**Comment by ajcvickers** [comment]
Notes from triage:
- Putting this on the backlog to make this work better
- We can consider warning before that, but implementing an effective warning that doesn't become spam may not be easy. (And will likely be a bit of a hack.)

**Comment by roji** [comment]
To add to the above, for people wanting to do dynamic tags, note that this cause changes in the actual resulting SQL, and therefore can have an impact on plan caching, which is important for performance. SQL Server specifically (apparently) strips leading comments away before caching SQL, but other databases don't (e.g. Npgsql's prepared statement feature for PostgreSQL).

In other words, even if we were to make this faster in EF itself, it would still cause performance issues on at least some databases - and is therefore not a good idea.

**Comment by roji** [comment]
> Then it's up to db engines to not make it their issue.

It's not an "issue" - that's how things usually work by-design; plan caching is keyed on the SQL, and barring some specific targeted exceptions (like SQL Server stripping out leading comments) that's how it's supposed to be. So even if the EF side is improved, by introducing dynamic tags you're causing some pretty bad performance issues; I'd wager that most people doing this are entirely unaware of the consequences, and wouldn't necessarily want to do this - so a throw-by-default warning (that can be suppressed) would make sense to me.

> IMO, EF could benefit from better distinguishing essentials from frills when caching queries. Out of curiosity, are similar issues lurking when caching Contains queries, each time generating different JSON?

The new OPENJSON translation for Contains sends a **parameterized** JSON array, so the SQL is always the same. This, in fact, was the main reason we moved to OPENJSON from the previous approach, where constants where embedded in the SQL (`WHERE x IN (1, 2, 3)`), which caused exactly the kind of query plan pollution discussed above (and was one of the most highly-voted issues in this repo).
